packages/wallet/src/redux/sagas/message-handler.ts

```diff
@@ -0,0 +1,28 @@
+import {put} from "redux-saga/effects";
+
+import * as actions from "../actions";
+import jrs, {RequestPayloadObject} from "jsonrpc-serializer";
+
+export function* messageHandler(jsonRpcMessage: string, fromDomain: string) {
+  const parsedMessage = jrs.deserialize(jsonRpcMessage);
+  switch (parsedMessage.type) {
+    case "notification":
+    case "success":
+      console.warn(`Received unexpected JSON-RPC message ${jsonRpcMessage}`);
+      break;
+    case "error":
+      throw new Error("TODO: Respond with error message");
+    case "request":
+      yield handleMessage(parsedMessage.payload as RequestPayloadObject, fromDomain);
+      break;
+  }
+}
+
+function* handleMessage(payload: jrs.RequestPayloadObject, domain: string) {
+  const {id} = payload;
+  switch (payload.method) {
+    case "GetAddress":
+      yield put(actions.addressRequest({id, domain}));
```

Comment:
It might be worth using this opportunity to start breaking out of the redux pattern. Especially in this case, where we just have a read operation, it seems like it would be possible to use the saga to read from the store and return a value, skipping the reducer etc. entirely.

Reply:
I've updated the saga to simply use a selector to get the `address` and then `fork` the messageSender so we skip the reducer entirely.